File: greedy/53.maximum-subarray.py
#
# @lc app=leetcode id=53 lang=python3
#
# [53] Maximum Subarray
#

# @lc code=start
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        res = float("-inf")
        sum = 0
        for i in range(len(nums)):
            sum += nums[i]
            res = sum if sum > res else res
            if sum <= 0:
                sum = 0

        return res
# @lc code=end

File: Greedy/53.maximum-subarray.java
/*
 * @lc app=leetcode id=53 lang=java
 *
 * [53] Maximum Subarray
 */

// @lc code=start
class Solution {
    public int maxSubArray(int[] nums) {
        int n = nums.length;
        int maxTotal = Integer.MIN_VALUE;
        int total = 0;

        for (int i = 0; i < n; i++) {
            total += nums[i];
            maxTotal = Math.max(maxTotal, total);
            total = Math.max(total, 0);
        }

        return maxTotal;
    }
}
// @lc code=end
